### PR TITLE
BUG: DataFrame.apply not adding a frequency if freq=None (#22150)

### LINKED ISSUE #22150: DataFrame.apply adds a frequency to a freq=None DatetimeIndex as a side-effect
#### Code Sample, a copy-pastable example if possible

```
import numpy as np, pandas as pd

def sudden_frequency(num_columns):
    index = pd.DatetimeIndex(["1950-06-30", "1952-10-24", "1953-05-29"])
    columns = list(range(num_columns))
    df = pd.DataFrame(np.random.random((len(index), num_columns)), index, columns)
    df.apply(lambda sr: sr)
    return index

for num_columns in range(5):
    print(num_columns, "--", sudden_frequency(num_columns))
```

Output:

```
0 -- DatetimeIndex(['1950-06-30', '1952-10-24', '1953-05-29'], dtype='datetime64[ns]', freq=None)
1 -- DatetimeIndex(['1950-06-30', '1952-10-24', '1953-05-29'], dtype='datetime64[ns]', freq=None)
2 -- DatetimeIndex(['1950-06-30', '1952-10-24', '1953-05-29'], dtype='datetime64[ns]', freq='WOM-4FRI')
3 -- DatetimeIndex(['1950-06-30', '1952-10-24', '1953-05-29'], dtype='datetime64[ns]', freq='WOM-4FRI')
4 -- DatetimeIndex(['1950-06-30', '1952-10-24', '1953-05-29'], dtype='datetime64[ns]', freq='WOM-4FRI')
```

#### Problem description

This particular index (found by [hypothesis](https://github.com/HypothesisWorks/hypothesis)) suddenly gains a frequency it is used in a DataFrame, with >= 2 columns, which goes on to call ".apply".

#### Expected Output

```
n -- DatetimeIndex(['1950-06-30', '1952-10-24', '1953-05-29'], dtype='datetime64[ns]', freq=None)
```

for all n.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-514.16.1.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: C
LOCALE: None.None

pandas: 0.23.3
pytest: 3.6.4
pip: 18.0
setuptools: 39.2.0
Cython: None
numpy: 1.15.0
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.5.0
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: 1.2.10
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by HannahFerch**
- [x] closes #22150
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by gfyoung** [comment]
@HannahFerch : Good start!  Will need a `whatsnew` and test(s) as well.

**Review by jreback — CHANGES_REQUESTED** [review]
comments

**Comment by mroeschke** [comment]
@HannahFerch ideally if you could include this test https://github.com/pandas-dev/pandas/issues/22150#issuecomment-410211224 in our testing suite, that'd be great.

I think `pandas/tests/frame/test_apply.py` should be an apt location for the test.

**Comment by HannahFerch** [comment]
@mroeschke Thanks, have tried to do as you suggested

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -628,7 +628,7 @@ Datetimelike
 - Bug in :meth:`DataFrame.eq` comparison against ``NaT`` incorrectly returning ``True`` or ``NaN`` (:issue:`15697`,:issue:`22163`)
 - Bug in :class:`DatetimeIndex` subtraction that incorrectly failed to raise `OverflowError` (:issue:`22492`, :issue:`22508`)
 - Bug in :class:`DatetimeIndex` incorrectly allowing indexing with ``Timedelta`` object (:issue:`20464`)
--
+- Bug in :class:`DatetimeIndex` where frequency was being set if original frequency was 'None' (:issue:`22150`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v0.24.0.txt]: Nit pick: Could you replace the quotes around `None` with double backticks? Like how Timedelta is in the line above.

File: pandas/tests/frame/test_apply.py [hunk 2]
```diff
@@ -1155,3 +1157,34 @@ def test_agg_cython_table_raises(self, df, func, expected, axis):
...
+
+    @given(index=indices(5), num_columns=integers(0, 5))
+    def test_frequency_is_original(self, index, num_columns):
+        # GH22150
+        original = index.copy()
+        df = DataFrame(True, index=index, columns=range(num_columns))
+        df.apply(lambda x: x)
+        assert index.freq == original.freq
+
+    def test_frequency_is_original_example(self):
```
Comment (mroeschke) [inline pandas/tests/frame/test_apply.py]: I don't think you need this test anymore since it should be covered by the test above.

**Comment by HannahFerch** [comment]
@mroeschke Done. Please let me know if anything else is needed from my side.

**Review by mroeschke — APPROVED** [review]
LGTM. @jreback

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @HannahFerch

**Pull request merged**